consensus/src/data_manager.rs

```diff
@@ -0,0 +1,235 @@
+                                            .replace(DataStatus::Requested(new_receivers));
+                                        return Err(BlockNotFound(block.id()));
+                                    }
+                                    Ok(Ok(data)) => {
+                                        debug!("QSE: got data, len {}", data.len());
+                                        vec_ret.push(data);
+                                    }
+                                    Ok(Err(e)) => {
+                                        debug!("QS: got error from receiver {:?}", e);
+                                        let new_receivers = self
+                                            .request_transactions(
+                                                proof_with_status.proofs.clone(),
+                                                LogicalTime::new(block.epoch(), block.round()),
+                                            )
+                                            .await;
+                                        // Could not get all data so requested again
+                                        proof_with_status
+                                            .status
+                                            .lock()
+                                            .replace(DataStatus::Requested(new_receivers));
+                                        return Err(e);
+                                    }
+                                }
+                            }
+                            let ret: Vec<SignedTransaction> =
+                                vec_ret.into_iter().flatten().collect();
+                            // execution asks for the data twice, so data is cached here for the second time.
+                            proof_with_status
+                                .status
+                                .lock()
+                                .replace(DataStatus::Cached(ret.clone()));
+                            Ok(ret)
+                        }
+                    }
+                }
+                _ => {
+                    // the Empty case is checked in the beginning
+                    warn!("should use QuorumStore");
+                    Ok(Vec::new())
+                }
+            }
+        } else {
+            match block.payload().unwrap() {
+                Payload::DirectMempool(txns) => Ok(txns.clone()),
+                _ => {
+                    // the Empty case is checked in the beginning
+                    warn!("should not use QuorumStore");
+                    Ok(Vec::new())
+                }
+            }
+        }
+    }
+
+    /// Since QuorumStore restarts every epoch, new_epoch updates the relevant communication information
+    #[allow(dead_code)]
+    pub fn new_epoch(
+        &self,
+        data_reader: Arc<BatchReader>,
+        quorum_store_wrapper_tx: Sender<WrapperCommand>,
+    ) {
+        if self.quorum_store_enabled.load(Ordering::Relaxed) {
```

Comment:
where do we set the bool? why is that not in the new_epoch function?

Reply:
I'm planning to just combine into a single `new_epoch` function in #5400 once this PR is landed.